worlds/mm2/rom.py

```diff
@@ -0,0 +1,415 @@
+        patch.write_byte(0x3)
+
+
+
+    from Utils import __version__
+    patch.name = bytearray(f'MM2{__version__.replace(".", "")[0:3]}_{world.player}_{world.multiworld.seed:11}\0',
+                           'utf8')[:21]
+    patch.name.extend([0] * (21 - len(patch.name)))
+    patch.write_bytes(0x3FFC0, patch.name)
+    deathlink_byte = world.options.death_link.value | (world.options.energy_link.value << 1)
+    patch.write_byte(0x3FFD5, deathlink_byte)
+
+    patch.write_bytes(0x3FFD8, world.world_version)
+
+    version_map = {
+        "0": 0x90,
+        "1": 0x91,
+        "2": 0x92,
+        "3": 0x93,
+        "4": 0x94,
+        "5": 0x95,
+        "6": 0x96,
+        "7": 0x97,
+        "8": 0x98,
+        "9": 0x99,
+        ".": 0xDC
+    }
+    patch.write_token(APTokenTypes.RLE, 0x36EE0, (11, 0))
+    patch.write_token(APTokenTypes.RLE, 0x36EEE, (25, 0))
+
+    # BY SILVRIS
+    patch.write_bytes(0x36EE0, [0xC2, 0xD9, 0xC0, 0xD3, 0xC9, 0xCC, 0xD6, 0xD2, 0xC9, 0xD3])
+    # ARCHIPELAGO x.x.x
+    patch.write_bytes(0x36EF2, [0xC1, 0xD2, 0xC3, 0xC8, 0xC9, 0xD0, 0xC5, 0xCC, 0xC1, 0xC7, 0xCF, 0xC0])
+    patch.write_bytes(0x36EFE, list(map(lambda c: version_map[c], __version__)))
+
+    patch.write_file("token_patch.bin", patch.get_token_binary())
+
+
+header = b"\x4E\x45\x53\x1A\x10\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+
+
+def read_headerless_nes_rom(rom: bytes) -> bytes:
+    if rom[:4] == b"NES\x1A":
+        return rom[16:]
+    else:
+        return rom
+
+
+def get_base_rom_bytes(file_name: str = "") -> bytes:
+    base_rom_bytes: Optional[bytes] = getattr(get_base_rom_bytes, "base_rom_bytes", None)
+    if not base_rom_bytes:
+        file_name = get_base_rom_path(file_name)
+        base_rom_bytes = read_headerless_nes_rom(bytes(open(file_name, "rb").read()))
+
+        basemd5 = hashlib.md5()
+        basemd5.update(base_rom_bytes)
+        if basemd5.hexdigest() == PROTEUSHASH:
+            base_rom_bytes = extract_mm2(base_rom_bytes)
+            basemd5 = hashlib.md5()
+            basemd5.update(base_rom_bytes)
+        if basemd5.hexdigest() not in {MM2LCHASH, MM2NESHASH, MM2VCHASH}:
+            print(basemd5.hexdigest())
+            raise Exception("Supplied Base Rom does not match known MD5 for US, LC, or US VC release. "
+                            "Get the correct game and version, then dump it")
+        headered_rom = bytearray(base_rom_bytes)
+        headered_rom[0:0] = header
+        setattr(get_base_rom_bytes, "base_rom_bytes", bytes(headered_rom))
+        return bytes(headered_rom)
+    return base_rom_bytes
+
+
+def get_base_rom_path(file_name: str = "") -> str:
+    options: settings.Settings = settings.get_settings()
+    if not file_name:
+        file_name = options["mm2_options"]["rom_file"]
+    if not os.path.exists(file_name):
+        file_name = Utils.user_path(file_name)
+    return file_name
+
+
+prg_offset = 0x8ED70
+prg_size = 0x40000
```

Comment:
PEP8 recommend that constants be in capital (https://peps.python.org/pep-0008/#constants).